python/pyspark/ml/image.py

```diff
@@ -0,0 +1,192 @@
+        return self._ocvTypes
+
+    @property
+    def imageFields(self):
+        """
+        Returns field names of image columns.
+
+        :rtype list: a list of field names.
+
+        .. versionadded:: 2.3.0
+        """
+
+        if self._imageFields is None:
+            ctx = SparkContext._active_spark_context
+            self._imageFields = list(ctx._jvm.org.apache.spark.ml.image.ImageSchema.imageFields())
+        return self._imageFields
+
+    @property
+    def undefinedImageType(self):
+        """
+        Returns the name of undefined image type for the invalid image.
+
+        .. versionadded:: 2.3.0
+        """
+
+        if self._undefinedImageType is None:
+            ctx = SparkContext._active_spark_context
+            self._undefinedImageType = \
+                ctx._jvm.org.apache.spark.ml.image.ImageSchema.undefinedImageType()
+        return self._undefinedImageType
+
+    def toNDArray(self, image):
+        """
+        Converts an image to a one-dimensional array.
+
+        :param image: The image to be converted
+        :rtype array: The image as a one-dimensional array
+
+        .. versionadded:: 2.3.0
+        """
+
+        height = image.height
+        width = image.width
+        nChannels = image.nChannels
+        return np.ndarray(
+            shape=(height, width, nChannels),
+            dtype=np.uint8,
+            buffer=image.data,
+            strides=(width * nChannels, nChannels, 1))
+
+    def toImage(self, array, origin=""):
+        """
+        Converts a one-dimensional array to a two-dimensional image.
+
+        :param array array: The array to convert to image
+        :param str origin: Path to the image
+        :rtype object: Two dimensional image
+
+        .. versionadded:: 2.3.0
+        """
+
+        if array.ndim != 3:
+            raise ValueError("Invalid array shape")
+        height, width, nChannels = array.shape
+        ocvTypes = ImageSchema.ocvTypes
+        if nChannels == 1:
+            mode = ocvTypes["CV_8UC1"]
+        elif nChannels == 3:
+            mode = ocvTypes["CV_8UC3"]
+        elif nChannels == 4:
+            mode = ocvTypes["CV_8UC4"]
+        else:
+            raise ValueError("Invalid number of channels")
+        data = bytearray(array.astype(dtype=np.uint8).ravel())
+        # Creating new Row with _create_row(), because Row(name = value, ... )
+        # orders fields by name, which conflicts with expected schema order
+        # when the new DataFrame is created by UDF
+        return _create_row(self.imageFields,
+                           [origin, height, width, nChannels, mode, data])
+
+    def readImages(self, path, recursive=False, numPartitions=0,
+                   dropImageFailures=False, sampleRatio=1.0):
+        """
+        Reads the directory of images from the local or remote source.
+
+        :param str path: Path to the image directory
+        :param SparkSession spark: The current spark session
```

Comment:
these param isn't present, instead you currently use _active_spark_context.

Reply:
good catch!  I added and then removed it in another commit.  Removed.